rfc/20240610-Tofu-Version-Compatibility.md

```diff
@@ -0,0 +1,109 @@
+
+> [!NOTE]
+> As we are unable to inspect the Hashicorp Terraform Repository in depth, all version compatibility must be based on Release Notes / Changelog Entries as well as user reports.
+
+## Proposed Solution
+
+As we introduce support for `.tofu` files, we open up the ability to add new language features without worrying as much about compatibility in shared modules. If we can differentiate between "OpenTofu Version" and "Terraform Version", the above scenarios become much easier.
+
+We must also consider "partial knowledge" of version information. Ideally the configuration would clearly define which terraform version and tofu version are required. If that is not the case, we will have to provide a "best guess" to fill in the blanks.
+
+As the [provider compatibility check is deprecated](https://github.com/hashicorp/terraform-plugin-framework/blob/8a09cef0bb892e2d90c89cad551ce12a959f1a2d/provider/configure.go#L28-L32) and the state's "terraform_version" field is not currently used in OpenTofu, we will not discuss them in this RFC.
+
+### User Documentation
+
+Currently the "software version" (both Terraform and Tofu) is defined in the config as `terraform -> required_version`. This represents a valid range of versions that the configuration is compatible with. When OpenTofu is compiled, the "software version" is baked into the binary and used for all version checks (as well as`tofu -version`).
+
+Example:
+```hcl
+terraform {
+    required_version = 1.7.4
+}
+```
+This configuration currently requires *both* Terraform 1.7.4 and OpenTofu 1.7.4 which are wildly different things.
+
+Let's look at a project or module which plans on supporting both tools (for the time being). They will start to introduce `.tofu` files as they start utilizing additional functionality in OpenTofu. How would they go about setting version requirements?
+
+```hcl
+# version.tf
+terraform {
+    required_version = 1.7.4
+}
+```
+```hcl
+# version.tofu
+tofu {
+    required_version = 1.6.5
+}
+```
+
+When loaded by:
+* Terraform, the `version.tofu` file would be ignored and the "Terraform Required Version" would be 1.7.4.
+* OpenTofu, the `version.tofu` file would be loaded instead of `version.tf` and the "OpenTofu Required Version" would be 1.6.5.
+
+
+What happens if only the `version.tf` file exists in this module with no tofu overwrite? This is a common case for modules who have not yet considered explicit OpenTofu migration.
+
+As a user, I would expect that the "Terraform Required Version" would have an equivalent "OpenTofu Required Version". This is not a 1-1 mapping, but is a reasonable stop-gap solution.
+
+As this is a not a 1-1 version mapping and a "best effort" solution, I would expect to see a warning:
+
+```shell
+$ tofu init
+...
+Warning: Using v1.7.x in 'terraform -> required_version' as equivalent to current tofu version 1.6.5!
```

Comment:
It's tricky though, we will need to communicate why the "required_version" constraint was not met in some scenarios.

We could split it into two messages, a warning for specifying `terraform -> required_version` and one that's more detailed when the constraint is not met.

Reply:
I think we need to reconsider the version mapping approach. It will become obsolete as soon as Terraform releases an incompatible feature.

After that, OpenTofu will still depend on module authors to update their configuration with alternative `.tofu` implementations.

Also, required version could be incompatible, but an actual configuration is fine to run with OpenTofu and Terraform (if it doesn't use incompatible features).